Date: Thu, 27 Feb 2020 19:26:32 +1100
Subject: [PATCH] added escape codes that change cursor color. 110-117 for dark
 8, 120-127 for bright 8, 130 for reset (make sure defaultcs is set to 256 in
 config.def.h)

echo -e "\033[110m" will change cursor to black
echo -e "\033[130m" will reset
---
 config.def.h |  2 +-
 st.c         | 20 ++++++++++++++++++++
 st.h         |  1 +
 3 files changed, 22 insertions(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 0e01717..a7bafac 100644
--- a/config.def.h
+++ b/config.def.h
@@ -118,7 +118,7 @@ static const char *colorname[] = {
  */
 unsigned int defaultfg = 7;
 unsigned int defaultbg = 0;
-static unsigned int defaultcs = 256;
+unsigned int defaultcs = 256;
 static unsigned int defaultrcs = 257;
 
 /*
diff --git a/st.c b/st.c
index b8e6077..3aae4b4 100644
--- a/st.c
+++ b/st.c
@@ -1442,6 +1442,14 @@ tsetattr(int *attr, int l)
 		case 49:
 			term.c.attr.bg = defaultbg;
 			break;
+		case 130: // reset cursor color
+            {
+                defaultcs = 256;
+                int cx = term.c.x;
+                xdrawcursor(cx, term.c.y, term.line[term.c.y][cx],
+                        term.ocx, term.ocy, term.line[term.ocy][term.ocx]);
+                break;
+            }
 		default:
 			if (BETWEEN(attr[i], 30, 37)) {
 				term.c.attr.fg = attr[i] - 30;
@@ -1451,6 +1459,18 @@ tsetattr(int *attr, int l)
 				term.c.attr.fg = attr[i] - 90 + 8;
 			} else if (BETWEEN(attr[i], 100, 107)) {
 				term.c.attr.bg = attr[i] - 100 + 8;
+			} else if (BETWEEN(attr[i], 110, 117)) { //cursor color dark 8
+				defaultcs = attr[i] - 110;
+                int cx = term.c.x;
+                xdrawcursor(cx, term.c.y, term.line[term.c.y][cx],
+                        term.ocx, term.ocy, term.line[term.ocy][term.ocx]);
+                break;
+			} else if (BETWEEN(attr[i], 120, 127)) { //cursor color bright 8
+				defaultcs = attr[i] - 120 + 8;
+                int cx = term.c.x;
+                xdrawcursor(cx, term.c.y, term.line[term.c.y][cx],
+                        term.ocx, term.ocy, term.line[term.ocy][term.ocx]);
+                break;
 			} else {
 				fprintf(stderr,
 					"erresc(default): gfx attr %d unknown\n",
diff --git a/st.h b/st.h
index 38c61c4..0eef297 100644
--- a/st.h
+++ b/st.h
@@ -120,3 +120,4 @@ extern char *termname;
 extern unsigned int tabspaces;
 extern unsigned int defaultfg;
 extern unsigned int defaultbg;
+extern unsigned int defaultcs;
-- 
2.25.1

